fix(kep-3633): redesign matchLabelKeys as matchLabelSelectors to introduce Operator field (#4001)
kubernetes/enhancements: keps/sig-scheduling/3633-matchlabelselectors-to-podaffinity/README.md

            values:
            - database
        topologyKey: topology.kubernetes.io/zone
        matchLabelSelectors: # ADDED
        - matchLabelKeys: pod-template-hash
          operator: In
```

#### Story 2

> When users want to achieve exclusive 1:1 job to domain placement in [kubernetes-sigs/jobset](https://github.com/kubernetes-sigs/jobset).

Reviewer: Great. their common key with the orchestrator are different value about Mpod&Npod.Can.we avoid this problem by specifying pod for foo？

> When users want to achieve exclusive 1:1 job to domain placement in [kubernetes-sigs/jobset](https://github.com/kubernetes-sigs/jobset).

Author: OK, I'll try to have another example here to explain the same usecase.

> When users want to achieve exclusive 1:1 job to domain placement in [kubernetes-sigs/jobset](https://github.com/kubernetes-sigs/jobset).

Author: As you suggested, I changed this example to be more general, not using jobset. PTAL.

I didn't follow your example because I feel like users can define LabelSelector of `foo=abc` / `foo=xyz` directly in Pod affinity / anti affinity. I think the point of the usecase of matchLabelKey is that "users cannot know the label value because, for example, the label value is applied from another system automatically (like jobset controller for `job-name`, the deployment controller for `pod-template-hash`, generated in the manifest manager, etc)".

Noone knows the job's name until it's created from JobSet, but each Job and Pod will get [`job-name`](https://github.com/kubernetes-sigs/jobset/blob/749a77c2647ac46b309b8227f66293e4116b97ed/api/v1alpha1/jobset_types.go#L26) label on it.

Users can use `job-name` label to ensure that the Pods with the same `job-name` will land on the same rack 
and Pods with other `job-name` won't labd on the same rack via `matchLabelSelectors`.

```yaml
apiVersion: jobset.x-k8s.io/v1alpha1
kind: JobSet
metadata:
  name: pytorch
  annotations:
    alpha.jobset.sigs.k8s.io/exclusive-topology: rack
spec:
  replicatedJobs:
    - name: workers
      template:
      ...
        affinity:
          podAffinity:      # ensures the pods of this job land on the same rack
              requiredDuringSchedulingIgnoredDuringExecution:
              - matchLabelSelectors:
                  - matchLabelKey: job-name
                    operator: In
                topologyKey: rack
            podAntiAffinity: # ensures only this job lands on the rack
              requiredDuringSchedulingIgnoredDuringExecution:
              - matchLabelSelectors:
                  - matchLabelKey: job-name

>                     operator: NotIn

Author: Could you provide how your proposal looks via example yaml or something?

>                     operator: NotIn

Reviewer 2: Here are some examples:

#### a set of Pods A doesn't want to co-exist with other set of Pods, but want the set of Pods A co-located

```yaml
    spec:
      affinity:
        podAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: pod-set
                operator: ExistsWithSameValue
            topologyKey: kubernetes.io/hostname
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: pod-set
                operator: ExistsWithDifferentValue
            topologyKey: kubernetes.io/hostname
```

- smooth rolling upgrade for PodAntiAffinity:
```yaml
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - pause
            topologyKey: kubernetes.io/hostname
          - labelSelector:
              matchExpressions:
              - key: pod-template-hash
                operator: ExistsWithSameValue
            topologyKey: kubernetes.io/hostname
```

>                     operator: NotIn

Reviewer 2: > @Huang-Wei I am also not following, which operator are you referring to?

(The term "operator" is misleading...) I meant [LabelSelectorOperator](https://github.com/kubernetes/kubernetes/blob/ea974280dc9948e31d426ee6366abff0957a5f82/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1215). And see ^^ for examples.

>                     operator: NotIn

Reviewer 3: I like @Huang-Wei's proposal better

>                     operator: NotIn

Reviewer 2: > Specifically, how would ExistsWithSameValue/ExistsWithDifferentValue be interpreted in the absence of an object associated with the selector?

It should be errored out, or downgraded to the `Exists` semantics. As `ExistsWithSameValue/ExistsWithDifferentValue` relies on the presence of the key in the incoming Pod (or other API object).

> we need a new match function to pass the label value I guess, unless we inject it in the value field, but who will do that?

It'd be the concrete feature impl. (in this proposal it's scheduling). Meanwhile the generic match function can enforce the semantics of `Exists`.

>                     operator: NotIn

Reviewer 4: > It should be errored out, or downgraded to the Exists semantics.

But where do we error out? perhaps in `LabelSelectorAsSelector`/`LabelSelectorAsMap` functions? 

> It'd be the concrete feature impl. (in this proposal it's scheduling). Meanwhile the generic match function can enforce the semantics of Exists.

I guess `LabelSelectorAsSelector`/`LabelSelectorAsMap` functions will need to be extended to allow passing in the value (or create new ones that accept it); `ExistsWithSameValue` translates to "In value" (just like the original MatchLabelKeys semantics); while ExistsWithDifferentValue translates to two requirements `Key exists` and `NotIn value`, with this, the Match func implementation doesn't need to change.

>  or downgraded to the Exists semantics.
> Meanwhile the generic match function can enforce the semantics of Exists.

I am not sure about this implicit behavior, it can be surprising.

>                     operator: NotIn

Reviewer 4: @liggitt any thoughts or reservations to extending the operator scope in label selector?

>                     operator: NotIn

Reviewer 5: Generally adding a new enum to a set is problematic in the API.  The readers in this case are the scheduler and any external management tools (descheduler, etc).  Does anyone else read this field for enforcement?  The kubelet?

>                     operator: NotIn

Author: > Does anyone else read this field for enforcement? The kubelet?

no one else in k/k will read it. (As you said, some external tools like descheduler or cluster autoscaler will read it though)

>                     operator: NotIn

Reviewer 4: but also, this is generally consumed via the labels pkg: https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apimachinery/pkg/labels; so if the new enums are handled there, most readers will just get them.

>                     operator: NotIn

Reviewer 3: Can you list the enum alternative in the "Alternatives" section of the KEP?

>                     operator: NotIn

Author: Sure, added this idea as an alternative for now. 
Who can give the final decision on whether the enum addition is allowed or not? (API reviewer?)

            values:
            - database
        topologyKey: topology.kubernetes.io/zone
        matchLabelSelectors: # ADDED

>         - matchLabelKeys: pod-template-hash

Reviewer 3: ```suggestion
        - key: pod-template-hash
```

          operator: In
```

#### Story 2

Let's say all Pods on each tenant get `tenant` label via a controller or a manifest management tool like Helm. 
And, the cluster admin now wants to achieve exclusive 1:1 tenant to domain placement.

> By applying the following affinity grobally with the mutating webhook, the cluster admin can ensure that the Pods with the same domain will land on the same domain, and Pods with other `tenant` won't land on the same domain. 

Reviewer 4: ```suggestion
By applying the following affinity globally using a mutating webhook, the cluster admin can ensure that the Pods from the same tenant will land on the same domain exclusively, meaning Pods from other `tenants` won't land on the same domain. 
```

            values:
            - database
        topologyKey: topology.kubernetes.io/zone
        matchLabelSelectors: # ADDED

>         - Key: pod-template-hash

Reviewer 3: ```suggestion
        - key: pod-template-hash
```

          operator: In
```

#### Story 2

Let's say all Pods on each tenant get `tenant` label via a controller or a manifest management tool like Helm. 

> And, the cluster admin now wants to achieve exclusive 1:1 tenant to domain placement.

Reviewer 2: Maybe emphasize a bit that the value of `tenant` is unknown when composing the yaml.

```suggestion
Although the value of `tenant` label is unknown when composing the workload's manifest, the cluster admin still wants to achieve exclusive 1:1 tenant to domain placement.
```

By applying the following affinity globally using a mutating webhook, the cluster admin can ensure that the Pods from the same tenant will land on the same domain exclusively, meaning Pods from other `tenants` won't land on the same domain. 

```yaml
affinity:
  podAffinity:      # ensures the pods of this tenant land on the same node pool
    requiredDuringSchedulingIgnoredDuringExecution:
      - matchLabelSelectors:
          - Key: tenant

>             operator: In

Reviewer 6: About the previous discussion of adding a new enum to label selector, I don't think we should do that, as existing clients are not prepared to handle that. Additionally, that new enum would make label selectors contextual, which would mean all code paths taking label selectors and returning matchers would then need to take label selectors AND a set of the labels of the containing object... that's a super invasive change I don't think we would ever successfully ripple out to all label selector consumers.

Having a separate spot in the API with key/operator fields that explicitly means "this label selector but with the value pulled from the pod's label" seems better to me than changing semantics of the existing label selector field.

I do wonder if it would make sense to have the API server actually populate the peer label selector field with the pod's label values on create. For example, creating this pod:
```yaml
metadata:
  labels:
    tenant: foo
affinity:
  podAffinity:      # ensures the pods of this tenant land on the same node pool
    requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: otherlabel
            operator: In
            values: ["othervalue"]
        matchLabelSelectors:
          - key: tenant
            operator: In
```
would result in this pod being persisted:
```diff
 metadata:
   labels:
     tenant: foo 
 affinity:
   podAffinity:      # ensures the pods of this tenant land on the same node pool
     requiredDuringSchedulingIgnoredDuringExecution:
       - labelSelector:
           matchExpressions:
           - key: otherlabel
             operator: In
             values: ["othervalue"]
+          - key: tenant
+            operator: In
+            values: ["foo"]
         matchLabelSelectors:
           - key: tenant
             operator: In
```

That would mean anything reading pods and understanding how they are scheduled wouldn't have to learn about this new field, and the podAffinity / podAntiAffinity fields would continue to contain all the scheduling info

Reviewer 3: +1 on doing this in apiserver.
But then we should do the same in https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades

Reviewer 6: (note that populating/merging with user specified fields on write in the server is usually not a great idea, but could be ok here specifically on pods since the field is immutable after create… also need to consider what would happen with an old client writing pods that didn't know about the new field and maybe copy the old value to the new pod in prepareforupdate)

Reviewer 4: That would simplify the implementation, but wouldn't users be surprised to see that duplicated in the labelSelector when reading back the spec?

Reviewer 6: would need to game it out for sure, but users would also be surprised to see their affinity/antiAffinity rules get ignored, which is what all existing API clients making scheduling decisions on pods would do until updated to understand this new field (alternate schedulers, autoscalers, etc), which might *never* happen

Author: Sure.

> But then we should do the same in https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades

It requires people to change the component where they need to pass the feature gate, doesn't it? 
If so, should we do this for PodAffinity only for now, and do this for PodTopologySpread after KEP-3243 is graduated to GA?

Reviewer 2: If "adding a new enum to label selector" has a huge blasting radius, and given we're gonna use API server for field population, is "matchLabelKeys" **_plus_** "notMatchLabelKeys" an alternative? so we don't need to restrict the operator validation on In/NotIn, and also the API field is similar with what we introduced in PodTopologySpread.

```go
matchLabelKeys: []  -> equivalent to a slice of selectors, all with operator In
notMatchLabelKeys: [] -> equivalent to a slice of selectors, all with operator NotIn
```

Author: Sorry, missed this comment.
But, I prefer the idea of keeping `matchLabelKeys` + adding `notMatchLabelKeys` to `matchLabelSelectors` because
> the API field is similar with what we introduced in PodTopologySpread.

makes both easier to understand instead of making a difference (matchLabelKeys vs matchLabelSelectors) between them to specify a similar thing.

@alculquicondor @ahg-g What do you think?

Reviewer 3: I'll defer to @ahg-g and @msau42

Author: @ahg-g @msau42 kindly ping. 🙏

Reviewer 4: If we can find a more intuitive name for `notMatchLabelKeys`, then I am onboard. Otherwise I would stick with the key+operator approach.

Author: - distinctLabelKeys
- ignoreSameValueLabelKeys
- ...? (My English skill can put out only them 😓

>       - matchLabelSelectors:
>           - key: tenant
>             operator: In
>         topologyKey: node-pool

Reviewer 2: ```suggestion
    - matchLabelSelectors:
        - key: tenant
          operator: In
      topologyKey: node-pool
```

  podAntiAffinity:  # ensures only Pods from this tenant lands on the same node pool
    requiredDuringSchedulingIgnoredDuringExecution:

>       - matchLabelSelectors:
>           - key: tenant
>             operator: NotIn
>       - labelSelector:
>           matchExpressions:
>           - key: tenant
>             operator: Exists
>         topologyKey: node-pool

Reviewer 2: ```suggestion
    - matchLabelSelectors:
        - key: tenant
          operator: NotIn
    - labelSelector:
        matchExpressions:
        - key: tenant
          operator: Exists
      topologyKey: node-pool
```

proposal will be implemented, this is the place to discuss them.
-->

A new optional field `MatchLabelSelectors` is introduced to `PodAffinityTerm`.

```go
type LabelSelectorOperator string

type MatchLabelSelector struct {

>   // Key is used to lookup value from the incoming pod labels, 

Reviewer 2: make the indents consistent? (you used tab below)

  // and that key-value label is merged with `LabelSelector`.
  // Key that doesn't exist in the incoming pod labels will be ignored. 
  Key  string
  // Operator defines how key-value, fetched via the above `Keys`, is merged into LabelSelector.
  // If Operator is `In`, `key in (value)` is merged with LabelSelector. 
  // If Operator is `NotIn`, `key notin (value)` is merged with LabelSelector. 
  //
  // +optional
  Operator       LabelSelectorOperator
}

type PodAffinityTerm struct {
	LabelSelector *metav1.LabelSelector
	Namespaces []string
	TopologyKey string
	NamespaceSelector *metav1.LabelSelector

	// MatchLabelSelectors is a set of pod label keys to select the group of existing pods 
  // which pods will be taken into consideration for the incoming pod's pod (anti) affinity. 

>   // The default value is empty.

Reviewer 2: Format the comments (to have same empty spaces)

	// +optional
	MatchLabelSelectors []strinMatchLabelSelectorg
}
```

When a Pod is created, kube-apiserver will obtain the labels from the pod 
labels by the key in `MatchLabelSelectors.Key`, and merge to `LabelSelector` of `PodAffinityTerm` depending on `Operator`:
- If Operator is `In`, `key in (value)` is merged with LabelSelector. 

> - If Operator is `NotIn`, `key notin (value)` is merged with LabelSelector. 

Reviewer 4: should we validate that only in and notin are supported? are there other relevant operators?

Reviewer 2: it'd be nice to add concrete examples here.

Author: > should we validate that only in and notin are supported?

+1. I don't think we will have supports for other fields. I added the mention that `in` and `notin` are valid.

proposal will be implemented, this is the place to discuss them.
-->

A new optional field `MatchLabelSelectors` is introduced to `PodAffinityTerm`.

```go

> type LabelSelectorOperator string

Reviewer 7: Can we define the allowed values? Are you planning only `In` and `NotIn` or sth else too?

type MatchLabelSelector struct {
  // Key is used to lookup value from the incoming pod labels, 
  // and that key-value label is merged with `LabelSelector`.
  // Key that doesn't exist in the incoming pod labels will be ignored. 
  Key  string
  // Operator defines how key-value, fetched via the above `Keys`, is merged into LabelSelector.
  // If Operator is `In`, `key in (value)` is merged with LabelSelector. 
  // If Operator is `NotIn`, `key notin (value)` is merged with LabelSelector. 
  //
  // +optional
  Operator       LabelSelectorOperator
}

type PodAffinityTerm struct {
  LabelSelector *metav1.LabelSelector
  Namespaces []string
  TopologyKey string
  NamespaceSelector *metav1.LabelSelector

  // MatchLabelSelectors is a set of pod label keys to select the group of existing pods 
  // which pods will be taken into consideration for the incoming pod's pod (anti) affinity. 
  // The default value is empty.
  // +optional

>   MatchLabelSelectors []strinMatchLabelSelectorg

Reviewer 7: nit: please fix the type

Author: Done in https://github.com/kubernetes/enhancements/pull/4069

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

> N/A

Author: Not related to matchLabelSelectors, but we decided not to have e2e test in https://github.com/kubernetes/kubernetes/pull/116065.

their Pod specs as it is.

To use this enhancement, users need to enable the feature gate (during this feature is in the alpha.),
and add `MatchLabelSelectors` on their PodAffinity/PodAntiAffinity.

**Downgrade**

> kube-apiserver will ignore `MatchLabelSelectors` in PodAffinity/PodAntiAffinity,

Reviewer 7: Actually - it will ignore (drop) setting this field for newly added objects.
But existing objects will continue to exist.

So I'm assuming that kube-scheduler will also be performing the actual computations under feature-gate, right?

Reviewer 7: This is also touching on
"How can this feature be enabled / disabled in a live cluster?" in PRR.

Author: > existing objects will continue to exist.

Ah, yes correct. This PR changes MatchLabelSelectors to be handled in kube-apiserver when Pod is created. 
So, existing objects will continue to have the labelselector made by MatchLabelSelectors.

What is the ideal action here? OK to leave the existing objects when turning off the feature gate, or we somehow need to ignore the labelselector made by MatchLabelSelectors (gonna be tricky though..).

#### Example

a set of Pods A doesn't want to co-exist with other set of Pods, but want the set of Pods A co-located

```yaml
spec:
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: pod-set
            operator: ExistsWithSameValue
        topologyKey: kubernetes.io/hostname
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: pod-set
            operator: ExistsWithDifferentValue
        topologyKey: kubernetes.io/hostname
```

smooth rolling upgrade for PodAntiAffinity:

```yaml
spec:
  affinity:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: app
            operator: In
            values:
            - pause
        topologyKey: kubernetes.io/hostname
      - labelSelector:
          matchExpressions:
          - key: pod-template-hash
            operator: ExistsWithSameValue
        topologyKey: kubernetes.io/hostname
```

> 

Reviewer 8: Can you add a brief summary of why this was not chosen?
